### ISSUE TITLE
(Socks) Proxy: use DNS of proxy

### ISSUE THREAD

**Issue by rezzorix** [louislam/uptime-kuma#2296]
### ⚠️ Please verify that this feature request has NOT been suggested before.

- [X] I checked and didn't find similar feature request

### 🏷️ Feature Request Type

Other

### 🔖 Feature description

After some testing it currently seems like the that if a proxy is configured, uptime-kuma is using its own local or defined DNS server even when a proxy is defined instead of using the proxy DNS.

### ✔️ Solution

Add option that when a proxy is configured, to also use conduct DNS resolution through this proxy.

**Comment by rezzorix** [comment]
@louislam this actually could also be considered a bug …. normally one would expect the proxy dns is used… no?

**Comment by jbrunner** [comment]
> @louislam this actually could also be considered a bug …. normally one would expect the proxy dns is used… no?

IMO this is not ab Bug. It should be possible to switch between socks5 and socks5h.

**Comment by rezzorix** [comment]
@jbrunner not sure if you understood what I was writing... it is not possible to switch.

My point was:
Currently defining a socks will still have the uptime-container resolver try to resolve a domain.
It is not possible to define the socks resolver - which by default the logic should be: IF socks configured THEN use socks resolver.

Example of usage:
Connecting via socks to check certain hostnames on a network only accessible via socks.
You would never be able to resolve any of the hostnames if the uptime-kuma resolver is used.

**Comment by jbrunner** [comment]
@rezzorix Maybe I missed your point. I'm new to uptime-kuma and missed the feature to resolve DNS trough socks. The proposed solution works for my HTTP checks as expected.

If you change the default behaviour of the existing "Socks v5" config, you might break existing implementations.

The proposed solution provides an additional option in the UI:

* "Socks v5" (Existing behaviour. Value is: socks5)
* "Socks v4 (+DNS)" (New. Value is: socks5h)

IMO using socks5 does not explicit mean to resolve dns through  socks. For example: curl does not resolve DNS trough socks by default:

    curl -x socks5://localhost:1008 http://example.com
    cur -x socks5h://localhost:1008 http://example.com

**Comment by rezzorix** [comment]
@jbrunner yes you are right, using socks5 does not explicit resolve dns through socks.

[The RFC1928](https://datatracker.ietf.org/doc/html/rfc1928) does not state anything about DNS/resolving at all.

On [Wikipedia](https://en.wikipedia.org/wiki/SOCKS) it states for Socks4 & 5 that 
>"The server should resolve the [domain name](https://en.wikipedia.org/wiki/Domain_name) and make connection to the destination host **if it can.**"

I have found 21+ year old bugzilla postings where exactly what I described is discussed:
[SOCKS5: DNS lookups (host resolving) should occur on proxy, not client side](https://bugzilla.mozilla.org/show_bug.cgi?id=134105).

If I read these threads right, then consensus is/was what I would have also expected:
That resolving should occur on proxy... but again, thats only 1 source and an old one too.

So agreed, rather than calling it a bug, it is probably better called a feature or specific configuration option that is missing in uptime-kuma.

Would be good if we could define within uptime-kuma that if proxy is used - the resolving also happens on proxy side.

For the how-to..... my coding skills are unfortunately not sufficient to tell 🙈
